internal/chain/reward.go

```diff
@@ -1,157 +1,312 @@
+					); err != nil {
+						slotError(err, payable)
+						return
+					}
+
+					subComm := subCommittee.FindCommitteeByID(cxLink.ShardID())
+					// _ are the missing signers, later for slashing
+					payableSigners, _, err := blockSigners(cxLink.Header(), subComm)
+					votingPower := votepower.Compute(subComm.Slots)
+
+					if err != nil {
+						slotError(err, payable)
+						return
+					}
+
+					for member := range payableSigners {
+						voter := votingPower.Voters[payableSigners[member].BlsPublicKey]
+						if !voter.IsHarmonyNode {
+							due := BlockRewardStakedCase.Mul(voter.EffectivePercent)
+							to := voter.EarningAccount
+							go func(signersDue numeric.Dec, addr common.Address, j int) {
+								payable <- slotPayable{
+									effective: signersDue,
+									payee:     addr,
+									nonce:     i * (i + j),
+									oops:      nil,
+								}
+							}(due, to, member)
+						}
+					}
+				}(i)
+			}
+
+			w.Wait()
+			resultsHandle := []slotPayable{}
+
+			for payThem := range payable {
+				resultsHandle = append(resultsHandle, payThem)
+			}
+
+			// Check if any errors
+			for payThem := range resultsHandle {
+				if err := resultsHandle[payThem].oops; err != nil {
+					return err
+				}
+			}
+			// Enforce order
+			sort.SliceStable(resultsHandle,
+				func(i, j int) bool {
+					return resultsHandle[i].nonce < resultsHandle[j].nonce
+				},
+			)
+
+			// Finally do the pay
+			for payThem := range resultsHandle {
+				state.AddBalance(
+					resultsHandle[payThem].payee,
+					resultsHandle[payThem].effective.TruncateInt(),
+				)
+			}
+		}
+		return nil
+	}
 
 	payable := []struct {
 		string
 		common.Address
 		*big.Int
 	}{}
 
+	_, signers, _, err := ballotResult(bc, header, header.ShardID())
+	if err != nil {
+		return err
+	}
+
 	totalAmount := rewarder.Award(
-		BlockReward, accounts, func(receipient common.Address, amount *big.Int) {
+		BlockReward, signers, func(receipient common.Address, amount *big.Int) {
 			payable = append(payable, struct {
 				string
 				common.Address
 				*big.Int
-			}{
-				common2.MustAddressToBech32(receipient), receipient, amount,
-			},
+			}{common2.MustAddressToBech32(receipient), receipient, amount},
 			)
 		},
 	)
 
-	if totalAmount.Cmp(BlockReward) != 0 {
+	if totalAmount.Equal(BlockReward) == false {
 		utils.Logger().Error().
 			Int64("block-reward", BlockReward.Int64()).
 			Int64("total-amount-paid-out", totalAmount.Int64()).
 			Msg("Total paid out was not equal to block-reward")
-		return errors.Wrapf(errPayoutNotEqualBlockReward, "payout "+totalAmount.String())
+		return errors.Wrapf(
+			errPayoutNotEqualBlockReward, "payout "+totalAmount.String(),
+		)
 	}
 
-	signers := make([]string, len(payable))
-
 	for i := range payable {
-		signers[i] = payable[i].string
 		state.AddBalance(payable[i].Address, payable[i].Int)
 	}
 
 	header.Logger(utils.Logger()).Debug().
-		Int("NumAccounts", len(accounts)).
+		Int("NumAccounts", len(payable)).
```

Comment:
It's better to print out for each shard separately. With numTotalVoters/numSigners/TotalRewards for each. Aggregated info is not helpful here. The totalAmount can be used to increase the TotalSupply record later.

Reply:
It see. this is the old logic. It's a bit confusing here.